text/0000-fn_name_macro.md

```diff
@@ -0,0 +1,61 @@
+- Feature Name: fn_name_macro
+- Start Date: 2019-11-14
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+<!-- https://github.com/rust-lang/rfcs/issues/1743 -->
+
+# Summary
+[summary]: #summary
+
+This RFC adds an additional macro, `function!`, to the `core` crate. When invoked, the macro expands to the name of the function that contains the call site.
+
+# Motivation
+[motivation]: #motivation
+
+This is a useful extension of Rust's existing debug reporting: `file!`, `line!`, `column!` and `module_path!`.
+
+For most people the name of the function is a much more immediate way to understand the context of a message than file name and line number.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For debug information about what's happening within the program there's several useful macros that you might use. One of them is the `function!` macro, which expands to the name of the current function. If used outside of a function it causes a compilation error.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Use of the `function!` macro expands to the compiler's internal name for the function. This will generally be the name that the user wrote into the file but in the case of a closure or similar it will be something like the function's name with a unique suffix.
+
+The exact text is considered "debug" information and not subject to Rust's stability guarantee.
+
+# Drawbacks
+[drawbacks]: #drawbacks
```

Comment:
It's another blow to incremental compilation (well it doesn't get worse than `column!()`, but this adds more...).

Reply:
Seems much less likely to cause such problems than column.

Reply:
Yeah (I said as much), but it should still be noted in the text to provide a more full examination of the trade-offs.

Reply:
How is incremental compilation particularly negatively affected? It seems like this would be a lot _less_ sensitive to file changes than `line!` and `column!` are.

Reply:
Yeah it is less sensitive than `line!()` certainly. It's probably the case that `function!()` or `item_path!()` is as bad as `module_path!()` is; only if you rename the item or parent module does the macro output change, so there's probably not a lot of impact on incremental after all.